cpp/src/arrow/compute/expression_test.cc

```diff
@@ -909,6 +909,39 @@ TEST(Expression, ExecuteCallWithNoArguments) {
   EXPECT_EQ(actual.length(), kCount);
 }
 
+TEST(Expression, ExecuteChunkedArray) {
```

Comment:
I think the fix is fine.
BTW, we should add one more test that the output data is also `ChunkedArray`, like

```c
  const int32_t length = 9;

  auto chunk_arr1 =
      ChunkedArrayFromJSON(int32(), {"[3, 1, 0, 3, null]", "[1, 2, 6, null]"});
  auto chunk_arr2 =
      ChunkedArrayFromJSON(int32(), {"[9, 0, 3, 1, null]", "[1, 8, 3, null]"});
  auto expr = call("min_element_wise", {field_ref("a"), field_ref("b")});
  auto input_schema = schema({field("a", int32()), field("b", int32())});
  ASSERT_OK_AND_ASSIGN(expr, expr.Bind(*input_schema));

  ExecBatch batch{{chunk_arr1, chunk_arr2}, length};
  ASSERT_OK_AND_ASSIGN(auto res, ExecuteScalarExpression(expr, batch));
  AssertDatumsEqual(
      res, ChunkedArrayFromJSON(int32(), {"[3, 0, 0, 1, null]", "[1, 2, 3, null]"}));
```
Because the comment https://github.com/apache/arrow/issues/41923#issuecomment-2143796671 mentioned that function kernel with `NullHandling::type::COMPUTED_NO_PREALLOCATE` will return chunked-array when the input is chunked-array.

And also, we should add chunked-array support to `MakeExecBatch`, because the expression system can handle chunked-array normally, so this can be opened as a separate issue include simplify these tests by `ExpectExecute`.

Reply:
> Because the comment #41923 (comment) mentioned that function kernel with NullHandling::type::COMPUTED_NO_PREALLOCATE will return chunked-array when the input is chunked-array.

I am not sure always expecting chunked-array as output should be tested. The implication from `COMPUTED_NO_PREALLOCATE` should be regarded as an unstable internal implementation choice.

> And also, we should add chunked-array support to MakeExecBatch, because the expression system can handle chunked-array normally, so this can be opened as a separate issue include simplify these tests by ExpectExecute.

Not really because `ExecBatch` should only contain scalars and simple arrays.

The right move is (probably) making `ExecuteScalarExpression` take `std::vector<Datum>` when a `ExecBatch` can't be built and pass that through the kernel matching logic that discerns between chunked and flat arrays.

There are many existing kernels taking `ExecBatch` assuming the input is either scalar or array, so you can't really transport more than that inside an `ExecBatch`.

Reply:
> The implication from `COMPUTED_NO_PREALLOCATE` should be regarded as an unstable internal implementation choice
Ah, in comparison, I don’t quite understand the meaning of the phrase `unstable internal implementation choice`.

Some functions have kernel output types of nested-types or dictionary-type, so they cannot pre-allocate validity-bitmap before actual execution, and can only set `NullHandling` to `COMPUTED_NO_PREALLOCATE`.

In this case, if the input of this function is chunked-array, according to the existing logic of `ScalarExecutor`, the output nested-types or dictionary-type must be chunked-array.

1. `preallocating_all_buffers_` must be `false` in this kind of kernel's execute path and go into `ExecuteNonSpans`
2. `ExecuteNonSpans` will emplace the every chunk's array-data result input listener.
 https://github.com/apache/arrow/blob/9ee6ea701e20d1b47934f977d87811624061d597/cpp/src/arrow/compute/exec.cc#L897
3. Listener have saved multi-arrays and return a chunked array in below logic
 https://github.com/apache/arrow/blob/9ee6ea701e20d1b47934f977d87811624061d597/cpp/src/arrow/compute/exec.cc#L814-L824


If If ExecBatch supports chunked-array, should such output also be tested (to ensure that the expression system works properly in this scenario)?

Reply:
I think the `is_scalar` is fixed, however, maybe this case contains more bugs if `chunked_array` is provided.

Reply:
@ZhangHuiGui I mean that a function with `NO_PREALLOCATE` could decide what to return during its execution, so a test that requires a specific choice seems overly rigid to me.

Reply:
I see. You're right. Thanks!